### PR TITLE
Decouple everything so we can support multiple targets and platforms

### PR DESCRIPTION
**Pull request by kbeckmann**
Would be cool to use the codebase and build for Thumbinator which is running on stm32f0xx, so need another sdk/hal.

We already know we want to support multiple targets, i.e. different boards with different features and/or pin configurations

### CODE REVIEW

**Comment by kbeckmann** [comment]
This is obviously not 100% done because it will conflict soooo much with your usb stuff artur. But what do you think of the general idea?

File: app/drivers/led.h [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <stdint.h>
```
Comment (arturo182) [inline app/drivers/led.h]: b -> i

**Comment by kbeckmann** [comment]
Yeah I did this on the plane cause I was a bit bored so I just brainstormed a bit.

My reasoning was that the _implementation_ of gpio, i2c, adc etc is highly _platform_ specific (stm32f0, stm32l4, efm?? etc). And separating targets from platforms seemed like the way to go. The names target and platform are perhaps not correct but felt like good names at the time.

**Comment by kbeckmann** [comment]
Any objections to the general structure? Should I continue and try to make this patch pr mergable?

**Comment by kbeckmann** [comment]
I think I've decoupled enough in order to create a decent baseline now to continue from. I haven't touched the USB stuff because I know it will lead to an insane amount of conflicts in your code. However, I need to change it sooner or later so we can port it to thumbinator...

File: Makefile [hunk 2]
```diff
@@ -4,45 +4,33 @@ CC      := $(CROSS_COMPILE)gcc
...
+	$(PLATFORM_SRCS) \
+
+TARGET_INCLUDES += \
+	targets \
+	targets/$(TARGET) \
+	platforms \
+	platforms/$(PLATFORM) \
+	$(PLATFORM_INCLUDES) \
+
+MIDDLEWARE_SRCS := 	\
```
Comment (arturo182) [inline Makefile]: Maybe += so we can add middlewares in targets and platforms. also no need for the spaces before \

File: app/drivers/uart.c [hunk 3]
```diff
@@ -3,7 +3,7 @@
 #include <queue.h>
 #include <task.h>
 
-#include "drivers/gpio.h"
+#include "platform/gpio.h"
```
Comment (arturo182) [inline app/drivers/uart.c]: after drivers, sort A->Z

File: app/platform/platform.h [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+#include "errors.h"
+
+#define EPLATFORM_INIT		(EPLATFORM_BASE + 0)
+
+err_t platform_init(void);
```
Comment (arturo182) [inline app/platform/platform.h]: init should be last

**Comment by kbeckmann** [comment]
Aside from the tiny details, is the structure good?

**Comment by kbeckmann** [comment]
btw fixed your comments

**Comment by kbeckmann** [comment]
can haz 👍?

**Comment by kbeckmann** [comment]
:(

**Comment by kbeckmann** [comment]
#87 contains this patch so this pr is not needed

**Pull request closed without merging**